Source/WebKit/WPEPlatform/wpe/WPESettings.cpp

```diff
@@ -0,0 +1,342 @@
+ *
+ * It is an error to call this function with a key that has already been registered.
+ */
+void wpe_settings_register(WPESettings* settingsObject, const char* key, const GVariantType* type, GVariant* defaultValue)
+{
+    g_return_if_fail(WPE_IS_SETTINGS(settingsObject));
+    g_return_if_fail(key && g_str_has_prefix(key, "/wpe-platform/") && !g_str_has_suffix(key, "/"));
+    g_return_if_fail(defaultValue);
+    g_return_if_fail(type);
+    g_return_if_fail(g_variant_type_equal(type, g_variant_get_type(defaultValue)));
+
+    if (settingsObject->priv->settings.contains(key)) {
+        g_warn_if_reached();
+        return;
+    }
+
+    settingsObject->priv->settings.add(key, SettingEntry {
+        nullptr,
+        g_variant_ref_sink(defaultValue),
+        GUniquePtr<GVariantType>(g_variant_type_copy(type))
+    });
+}
+
+/**
+ * wpe_settings_load_from_keyfile:
+ * @settings: a #WPESettings
+ * @keyfile: the keyfile to load settings from
+ * @error: return location for a #GError, or %NULL
+ *
+ * Loads settings from a keyfile.
+ *
+ * Only groups named `wpe-platform` or with the prefix `wpe-platform/` will be considered.
+ * All keys under these groups must be registered with wpe_settings_register().
+ *
+ * If a value is already set it will be overwritten.
+ *
+ * All keys loaded are expected to be formatted as a GVariant.
+ *
+ * [signal@WPESettings::changed] will be emitted for each key that is loaded with a *new* value.
+ *
+ * Returns: %TRUE if the settings were loaded successfully, %FALSE otherwise.
+ */
+gboolean wpe_settings_load_from_keyfile(WPESettings* settingsObject, GKeyFile* keyFile, GError** error)
+{
+    g_return_val_if_fail(WPE_IS_SETTINGS(settingsObject), FALSE);
+    g_return_val_if_fail(!error || !*error, FALSE);
+    g_return_val_if_fail(keyFile, FALSE);
+
+    GUniquePtr<char*> groups(g_key_file_get_groups(keyFile, nullptr));
+    for (uint i = 0; groups.get()[i]; i++) {
+        const char* group = groups.get()[i];
+
+        if (!g_str_has_prefix(group, "wpe-platform/") && strcmp(group, "wpe-platform"))
+            continue;
+
+        GUniquePtr<char*> keys(g_key_file_get_keys(keyFile, group, nullptr, nullptr));
+
+        for (uint k = 0; keys.get()[k]; k++) {
+            const char* key = keys.get()[k];
+            GUniquePtr<char> value(g_key_file_get_value(keyFile, group, key, nullptr));
+            if (!value)
+                continue;
+
+            auto path = makeKeyPath(group, key);
+            auto iter = settingsObject->priv->settings.find(path.data());
+            if (iter == settingsObject->priv->settings.end()) {
+                g_set_error(error, WPE_SETTINGS_ERROR, WPE_SETTINGS_ERROR_NOT_REGISTERED, "Key %s not registered", path.data());
+                return FALSE;
+            }
+
+            GRefPtr<GVariant> parsedValue = adoptGRef(g_variant_parse(iter->value.type.get(), value.get(), nullptr, nullptr, error));
+            if (!parsedValue)
+                return FALSE;
```

Comment:
`g_variant_parse()` sets `error` itself.

Reply:
Right, but since we have a specific error quark, I think we should use it for any error in this api. So, I would build a `WPE_SETTINGS_ERROR` with the info from the variant error in the message.